toc/rfc/rfc-draft-ipv6-dual-stack-for-cf.md

```diff
@@ -0,0 +1,200 @@
+
+The CNI wrapper plugin must be extended to support IPv6. Currently, it creates default iptables rules (`netin-*`, `netout-*` chains) with default `REJECT` rules and `ALLOW` rules for DNS and "host services." In deployments without vxlan-policy-agent enabled, it also manages iptables egress rules. The enhancement requires adding logic to handle IPv6 rules using ip6tables.
+
+Additionally, the VXLAN Policy Agent must be updated to support IPv6 Address Security Groups (ASGs) received from the Policy Server.
+
+### Cloud Controller
+
+Changes to the cloud controller are primarily in the following areas:
+
+1. Support forwarding and retrieving of IPv6 addresses in any records that may have IP addresses
+2. Support binding to IPv4 and IPv6 for the CAPI server, depending on the VM’s configuration
+3. Support of IPv6 in metadata, e.g. Application Security Groups
+
+### Cloud Foundry CLI
+
+The Cloud Foundry CLI shows the application workload’s addresses in various places. Those places need to be extended to support the display of IPv6 addresses, as retrieved via the CF API:
+
+1. IPv6 support for configuring application security groups
+2. …
+
+### Networking
+
+Network Policies that are stored in [the policy-server-api which is part of cf-networking-release](https://github.com/cloudfoundry/cf-networking-release/blob/develop/docs/08-policy-server-api.md).
+
+Network policies may reference IP address ranges as destination. Destinations are stored as strings. The underlying IP ranges are also [strings with a start and end address](https://github.com/cloudfoundry/cf-networking-release/blob/aa25a3f96dc21e8a3e25db10f741e4f678c78464/src/code.cloudfoundry.org/policy-server/api/api.go#L37-L49). The data is ultimately serialized JSON that is stored in a database column. As IPv6 addresses are longer than IPv4 addresses, we might need to bump the table size.
+
+### Routing
+
+The largest impact on routing is Gorouter and in tcp-router via and routing-api. They all consume endpoint information provided by route-registrar or route-emitter. Such endpoint information usually points to an IP address and port.
+
+Endpoint information must support IPv6 address + port information in addition to IPv4 addresses.
+
+![Routing architecture](https://docs.cloudfoundry.org/concepts/images/cf-routing-architecture.png)
+
+#### Gorouter
+
+Gorouter is largely based on the Golang implementation for HTTP request processing and routing. This implementation is capable of IPv6 address resolution and communication. Gorouter’s route registry and its endpoint handling are also well contained
+
+For the scope of this RFC, testing and fixes for small unforeseen issues are considered, but no direct work is expected for supporting IPv6 targets.
+
+#### TCP-Router and Routing API
+
+routing-api and tcp-router need to be verified, and if necessary extended, to support IPv6 addresses in addition to IPv4 addresses.
+
+## Other Topics
+
```

Comment:
I think another thing probably to mention is that all components that have API (Cloud Controller, Policy Server, UAA, Credhub, BBS, Rep, Mysql) should be able to accept IPv6 connections. It is probably already managed by the http modules they use, unless some explicitly listen on 127.0.0.1.

Reply:
Also, BOSH DNS should resolve IPv6 addresses depending on the client protocol.

Reply:
Another thing, CATS should have end to end tests for IPv6 (ASG rules, App listening on IPv6, etc)

Reply:
I've added the point about "listeners" generally being IPv6 / dual stack capable, also including BOSH DNS. 

CATs and bbl (thanks @jochenehret) are now mentioned as well.